content/blog/2022-05-17-object-oriented-features/index.md

```diff
@@ -0,0 +1,210 @@
+
+This class needs no conversion to IR because `ClassDefn` will do it all by entering the body.
+
+## `ClassDecl`
+
+This class needs no conversion to IR. At this point, all symbols are collected and type check has occurred. There is no use for `ClassDecl` anymore.
+
+## `FunctionDefn`
+
+FunctionDefns have a field isMethod. When we convert a method FunctionDefn to IR we must treat it differently than a normal function. Specifically, the ABI naming convention is different for functions and methods. We will discuss this convention below. But because of the difference between functions and methods, we do not convert to IR for methods and let the owning ClassDefn do the conversion. We refrain from allowing a method FunctionDefn from doing a conversion and discarding it because the conversion mutates state in the NodeToIRConverter and we would have to undo it later.
+
+## `ClassDefn`
+
+Like we just said above, `ClassDefn` is responsible for converting the `FunctionDefns` to IR using the correct method naming convention. It is also responsible for generating a dispatch vector for the class. This generation is quite involved and we will discuss in depth below.
+
+## `FieldAccess`
+
+This node converts itself by looking up the index of the field in the object representation and getting the offset into that array.
+
+## `MethodCall`
+
+We lookup the function in the dispatch vector and generate the same IR as a normal `FExpr` except that a `FExpr` uses a `IRName` but we are using a `IRMem` that contains the name from the DV lookup. We don't need to worry about passing the object to the function because the `ThisAugmenter` already took care of this for us`.
+
+## `New`
+
+We must allocate the new object and then call the initialization function. First we build the dispatch vector and get a pointer to it. Then we convert all the fields to IR. Then we create an array that contains this pointer followed by the fields. 
+
+## `NodeToIRConverter`
+
+We introduces a few new fields for bookkeeping:
+
+1. `objectMethodEncodings` to store mangled function names. This allows us to store a mapping from method name to the actual ABI encoding for that function.T he encoding is the same as the encoding defined for a function or procedure in the Xi ABI except that the function name is proceeded by the class name and an underscore. See the section on DispatchVectorClassResolver for more information.
+2. `classNameToNumFields` represents a mapping from class name to the number of fields that class has
+3. `classNameToFieldNameToIndex` represents a mapping from class name to the index of the field in the object representation.
+
+This class is also responsible for building the dispatch vector and getting the method symbol out of a dispatch vector.
+
+# Dispatch Vectors
+
+Dispatch Vectors were the biggest focus during this project. According to Professor Myers, there are [3 goals](https://www.cs.cornell.edu/courses/cs4120/2022sp/notes.html?id=objects) we had:
```

Comment:
I also think they were the most interesting part! Is there any way to organize the blog post so this interesting part gets highlighted as much as possible, even if it means removing/demoting a bunch of the earlier stuff that is less interesting?